src/sentry/api/authentication.py

```diff
@@ -550,3 +551,91 @@ def authenticate_token(self, request: Request, token: str) -> tuple[Any, Any]:
         sentry_sdk.get_isolation_scope().set_tag("rpc_auth", True)
 
         return (AnonymousUser(), token)
+
+
+def compare_service_signature(
+    url: str,
+    body: bytes,
+    signature: str,
+    shared_secret_setting: list[str] | None,
+    service_name: str,
+) -> bool:
+    """
+    Generic function to compare request data + signature signed by one of the shared secrets.
+
+    Once a key has been able to validate the signature other keys will
+    not be attempted. We should only have multiple keys during key rotations.
+
+    Args:
+        url: The request URL path
+        body: The request body
+        signature: The signature to validate
+        shared_secret_setting: List of shared secrets from settings
+        service_name: Name of the service for logging (e.g., "Seer", "Launchpad")
+    """
+
+    if not shared_secret_setting:
+        raise RpcAuthenticationSetupException(
+            f"Cannot validate {service_name} RPC request signatures without shared secret"
+        )
+
+    if not signature.startswith("rpc0:"):
+        logger.error("%s RPC signature validation failed: invalid signature prefix", service_name)
+        return False
+
+    if not body:
+        logger.error("%s RPC signature validation failed: no body", service_name)
+        return False
+
+    try:
+        # We aren't using the version bits currently.
+        _, signature_data = signature.split(":", 2)
+
+        signature_input = body
+
+        for key in shared_secret_setting:
+            computed = hmac.new(key.encode(), signature_input, hashlib.sha256).hexdigest()
+            is_valid = constant_time_compare(computed.encode(), signature_data.encode())
+            if is_valid:
+                return True
+    except Exception:
+        logger.exception("%s RPC signature validation failed", service_name)
+        return False
+
+    logger.error("%s RPC signature validation failed", service_name)
+
+    return False
+
+
+class ServiceRpcSignatureAuthentication(StandardAuthentication):
```

Comment:
rip of https://github.com/getsentry/sentry/blob/master/src/sentry/api/endpoints/seer_rpc.py#L101